### PR TITLE
HP fortify issue Fix

### PR DESCRIPTION
**Pull request by Rajsmit**
Fixed some of the security issue

1. Code correctness :- Constructor invokes overridable function :-
CoapServer.java, BlockOption.java, Option.java, RemoteEndpoint.java,

2. Insecure Randomness :- MessageIdTracker.java, ReliabilityLayer.java

3. Redundent Null Check :- CoapServer.java, UdpMatcher.java

4. Object Model violation Just one equals() and HashCode() defined :-
BlockOption.java

5. Code Correctness :- Errorneous string Compare in Switch case.
Deduplicator.java

Signed-off-by: Rajesh Kumar <rajsmit@gmail.com>

### CODE REVIEW

File: californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java [hunk 1]
```diff
@@ -147,7 +147,7 @@ public CoapServer(final NetworkConfig config, final int... ports) {
 		}
 		
 		// resources
-		this.root = createRoot();
+		this.root = new RootResource();
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java]: This effectively removes the possibility to use another root resource by extending `CoapServer`.

File: californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java [hunk 2]
```diff
@@ -157,13 +157,17 @@ public CoapServer(final NetworkConfig config, final int... ports) {
...
 		// endpoints
 		this.endpoints = new ArrayList<>();
-		// sets the central thread pool for the protocol stage over all endpoints
-		this.executor = Executors.newScheduledThreadPool(//
-				config.getInt(NetworkConfig.Keys.PROTOCOL_STAGE_THREAD_COUNT), //
-				new Utils.NamedThreadFactory("CoapServer#")); //$NON-NLS-1$
-		// create endpoint for each port
-		for (int port : ports) {
-			addEndpoint(new CoapEndpoint(port, this.config));
+		if(null != config) {
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java]: The solution to this problem is not checking for `null` but using `this.config` instead (it has been initialized to a non-null value in the code above).

File: californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java [hunk 3]
```diff
@@ -305,7 +309,7 @@ public MessageDeliverer getMessageDeliverer() {
 	 * @param endpoint the endpoint to add
 	 */
 	@Override
-	public void addEndpoint(final Endpoint endpoint) {
+	public final void addEndpoint(final Endpoint endpoint) {
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java]: why do we need to make this method final and none of the others?

File: californium-core/src/main/java/org/eclipse/californium/core/coap/BlockOption.java [hunk 4]
```diff
@@ -235,6 +235,14 @@ public boolean equals(Object o) {
 		return szx == block.szx && num == block.num && m == block.m;
 	}
 
+	@Override
+	public int hashCode() {
+		int result = szx;
+		result = 31 * result + (m ? 1 : 0);
+		result = 31 * result + num;
+		return result;
+	}
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/coap/BlockOption.java]: nice catch

File: californium-core/src/main/java/org/eclipse/californium/core/coap/Option.java [hunk 5]
```diff
@@ -199,7 +199,7 @@ public long getLongValue() {
 	 *
 	 * @param value the new value
 	 */
-	public void setValue(byte[] value) {
+	public final void setValue(byte[] value) {
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/coap/Option.java]: Why not make the whole class final?

File: californium-core/src/main/java/org/eclipse/californium/core/network/MessageIdTracker.java [hunk 6]
```diff
@@ -57,7 +57,7 @@ public MessageIdTracker(final NetworkConfig config) {
 		exchangeLifetime = config.getLong(NetworkConfig.Keys.EXCHANGE_LIFETIME);
 		boolean useRandomFirstMID = config.getBoolean(NetworkConfig.Keys.USE_RANDOM_MID_START);
 		if (useRandomFirstMID) {
-			counter = new AtomicInteger(new Random().nextInt(1 << 10));
+			counter = new AtomicInteger(new SecureRandom().nextInt(1 << 10));
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/network/MessageIdTracker.java]: There is no need to use a cryptographically strong random in order to create the initial counter value. The counter will be incremented sequentially only afterwards ...

File: californium-core/src/main/java/org/eclipse/californium/core/observe/ObserveRelation.java [hunk 7]
```diff
@@ -163,7 +163,7 @@ public Exchange getExchange() {
 	 *
 	 * @return the source address
 	 */
-	public InetSocketAddress getSource() {
+	public final InetSocketAddress getSource() {
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/observe/ObserveRelation.java]: Why do we need to make this method final but not the rest?

File: californium-core/src/main/java/org/eclipse/californium/core/server/resources/ConcurrentCoapResource.java [hunk 8]
```diff
@@ -125,7 +125,7 @@ public ConcurrentCoapResource(String name, int threads) {
 	 * 
 	 * @param executor the executor service
 	 */
-	public void setExecutor(ExecutorService executor) {
+	public final void setExecutor(ExecutorService executor) {
```
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/server/resources/ConcurrentCoapResource.java]: same here, why only these two methods?

File: californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java (hunk 1 above)
Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java]: Maybe it is more about providing means for subclasses to provide a different *banner* for `GET /` requests? That could be easily achieved ...

File: californium-core/src/main/java/org/eclipse/californium/core/coap/BlockOption.java (hunk 4 above)
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/coap/BlockOption.java]: I suppose the idea is to say that blockOption is now immutable ? 
- Should we still need setter (as there are private now)?
- Should we not define szx, num as final ?

File: californium-core/src/main/java/org/eclipse/californium/core/network/deduplication/DeduplicatorFactory.java [hunk 9]
```diff
@@ -69,14 +69,14 @@ public static synchronized void setDeduplicatorFactory(DeduplicatorFactory facto
 	public Deduplicator createDeduplicator(final NetworkConfig config) {
 
 		String type = config.getString(NetworkConfig.Keys.DEDUPLICATOR, NetworkConfig.Keys.NO_DEDUPLICATOR);
-		switch(type) {
-		case NetworkConfig.Keys.DEDUPLICATOR_MARK_AND_SWEEP:
+		if (type.equals(NetworkConfig.Keys.DEDUPLICATOR_MARK_AND_SWEEP)) {
```
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/network/deduplication/DeduplicatorFactory.java]: Why replacing switch by if/else if ?
https://docs.oracle.com/javase/7/docs/technotes/guides/language/strings-switch.html

File: californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java (hunk 1 above)
Comment (sbernard31) [inline californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java]: Why should we limit  custom behavior to  `GET /` request only ?

E.g. For  `Bootstrap delete` in LWM2M, you need a custom behavior on `DETELE /`  :)

Comment (sophokles73) [inline californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java]: I didn't want to limit the possibilities, I was just suspecting that people might not want the default response with the `CoAP RFC Californium` bla bla bla ;-)

Comment (Rajsmit) [inline californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java]: Right now it is very urgent for us to fix it. We can still live with it. For now we can park this until we find a better solution for this.

File: californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java (hunk 2 above)
Comment (Rajsmit) [inline californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java]: I Agree, I will change it to use  this.config instead config and remove null check as this.config is non-null.

File: californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java (hunk 3 above)
Comment (Rajsmit) [inline californium-core/src/main/java/org/eclipse/californium/core/CoapServer.java]: This is same as point 1 so lets park it for now until we have more structured solution.

File: californium-core/src/main/java/org/eclipse/californium/core/network/deduplication/DeduplicatorFactory.java (hunk 9 above)
Comment (Rajsmit) [inline californium-core/src/main/java/org/eclipse/californium/core/network/deduplication/DeduplicatorFactory.java]: I agree with you while using jdk7 and above it is not an issue.

File: californium-core/src/main/java/org/eclipse/californium/core/network/MessageIdTracker.java (hunk 6 above)
Comment (Rajsmit) [inline californium-core/src/main/java/org/eclipse/californium/core/network/MessageIdTracker.java]: I Agree with you. I will remove it.

File: californium-core/src/main/java/org/eclipse/californium/core/coap/Option.java (hunk 5 above)
Comment (Rajsmit) [inline californium-core/src/main/java/org/eclipse/californium/core/coap/Option.java]: This is same as point 1 so lets park it for now until we have more structured solution.

File: californium-core/src/main/java/org/eclipse/californium/core/observe/ObserveRelation.java (hunk 7 above)
Comment (Rajsmit) [inline californium-core/src/main/java/org/eclipse/californium/core/observe/ObserveRelation.java]: This is same as point 1 so lets park it for now until we have more structured solution.

**Comment by sbernard31** [comment]
sounds good to me now :)

**Pull request merged**